mp4parse/src/lib.rs

```diff
@@ -1503,6 +1508,178 @@ fn read_infe<T: Read>(src: &mut BMFFBox<T>) -> Result<ItemInfoEntry> {
     Ok(ItemInfoEntry { item_id, item_type })
 }
 
+fn read_iref<T: Read>(src: &mut BMFFBox<T>) -> Result<TryVec<ItemReferenceEntry>> {
+    let mut entries = TryVec::new();
+    let version = read_fullbox_version_no_flags(src)?;
+    if version > 1 {
+        return Err(Error::Unsupported("iref version"));
+    }
+
+    let mut iter = src.box_iter();
+    while let Some(mut b) = iter.next_box()? {
+        let from_item_id = if version == 0 {
+            be_u16(&mut b)? as u32
+        } else {
+            be_u32(&mut b)?
+        };
+        let item_count = be_u16(&mut b)?;
+        for _ in 0..item_count {
+            let to_item_id = if version == 0 {
+                be_u16(&mut b)? as u32
+            } else {
+                be_u32(&mut b)?
+            };
+            entries.push(ItemReferenceEntry {
+                item_type: b.head.name.into(),
+                from_item_id,
+                to_item_id,
+            })?;
+        }
+    }
+    Ok(entries)
+}
+
+fn read_iprp<T: Read>(src: &mut BMFFBox<T>) -> Result<TryVec<AssociatedProperty>> {
+    let mut iter = src.box_iter();
+    let mut properties = TryVec::new();
+    let mut associations = TryVec::new();
+
+    while let Some(mut b) = iter.next_box()? {
+        match b.head.name {
+            BoxType::ItemPropertyContainerBox => {
+                properties = read_ipco(&mut b)?;
+            }
+            BoxType::ItemPropertyAssociationBox => {
+                associations = read_ipma(&mut b)?;
+            }
+            _ => return Err(Error::InvalidData("unexpected ipco child")),
+        }
+    }
```

Comment:
I see you prefer strictly validating parser, rather than merely parser that can read spec-compliant files.

In this case I think code could be simpler if validity was checked separately after parsing.

Reply:
In this case, how would you recommend doing the validity check after parsing?

On a more general note, you're right, I do prefer a strictly validating parser. Part of that is just my personal preference, but I think the better argument for it in this case is that AVIF is a new format, and therefore requires new software to be written to produce AVIF files. Much of that will likely rely on existing code for generating ISOBMFF-, MIAF- and HEIF-implementing files, but some will be written from scratch so I think the risk of code which generates files that violate the specs is highest now. Likewise, this is the time that a strict parser which generates errors is most likely to result in those errors being recognized and fixed. But also, in trying to adhere to the spec strictly, I find it's more likely to expose situations where I (or other authors of the parser) have misunderstood the spec. Finally, if technically invalid files become pervasive, it's always possible to make the parser more permissive in the future, but the opposite is rarely acceptable. I'm reminded of the general situation with early HTML where a policy of permissive parsing led to different behavior across browsers and nightmarish complexity in an effort to maintain "quirky", nonstandard behavior that websites had grown to rely on.

All that said, I like working with people who have different perspectives from my own. It helps me to explore approaches I wouldn't have considered and generate better solutions. And especially with specs like ISOBMFF the stricter implementation can end up being less elegant, so I want to be mindful of the overall readability of the code as well as it's fidelity to the spec.